File: resources/androidstudio/app/src/main/java/com/nativephp/mobile/ui/nativerender/NativeElementObservationRegistry.kt
package com.nativephp.mobile.ui.nativerender

import android.util.Log
import org.json.JSONArray
import org.json.JSONObject
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicInteger

/**
 * Opt-in observation seam for tooling that consumes rendered trees or native
 * interaction events. Payloads are serialized only while observers exist.
 */
object NativeElementObservationRegistry {
    enum class Stream { TREES, EVENTS }

    data class Subscription internal constructor(
        internal val stream: Stream,
        internal val id: Int,
    )

    private const val TAG = "NativeElementObservers"
    private val sequence = AtomicInteger(0)
    private val treeLock = Any()
    private val treeObservers = linkedMapOf<Int, (String) -> Unit>()
    private val eventObservers = ConcurrentHashMap<Int, (String) -> Unit>()
    private var latestTree: NativeUITree? = null

    fun observeTrees(observer: (String) -> Unit): Subscription {
        val id = sequence.incrementAndGet()
        val replay = synchronized(treeLock) {
            treeObservers[id] = observer
            latestTree
        }
        replay?.let { tree -> serializeTree(tree)?.let { notify(listOf(observer), it, "tree") } }
        return Subscription(Stream.TREES, id)
    }

    fun observeEvents(observer: (String) -> Unit): Subscription =
        subscribeEvents(observer)

    fun unsubscribe(subscription: Subscription) {
        when (subscription.stream) {
            Stream.TREES -> synchronized(treeLock) { treeObservers.remove(subscription.id) }
            Stream.EVENTS -> eventObservers.remove(subscription.id)
        }
    }

    internal fun publishTree(tree: NativeUITree) {
        val observers = synchronized(treeLock) {
            latestTree = tree
            treeObservers.values.toList()
        }
        if (observers.isEmpty()) return
        val json = serializeTree(tree) ?: return
        notify(observers, json, "tree")
    }

    internal fun publishEvent(type: Int, callbackId: Int, nodeId: Int) {
        if (type > EventType.SHEET_DISMISS) return
        if (eventObservers.isEmpty()) return
        val json = runCatching {
            JSONObject()
                .put("type", type)
                .put("callback_id", callbackId)
                .put("node_id", nodeId.toLong() and 0xffffffffL)
                .put("timestamp", System.currentTimeMillis())
                .toString()
        }.getOrNull() ?: return
        notify(eventObservers.values, json, "event")
    }

    private fun subscribeEvents(observer: (String) -> Unit): Subscription {
        val id = sequence.incrementAndGet()
        eventObservers[id] = observer
        return Subscription(Stream.EVENTS, id)
    }

    private fun notify(
        observers: Collection<(String) -> Unit>,
        json: String,
        label: String,
    ) {
        for (observer in observers) {
            try {
                observer(json)
            } catch (t: Throwable) {
                Log.w(TAG, "$label observer error: ${t.message}")
            }
        }
    }

    internal fun serializeTree(tree: NativeUITree): String? = runCatching {
        JSONObject()
            .put("version", tree.version)
            .put("callback_count", tree.callbackCount)
            .put("root", serializeNode(tree.root))
            .toString()
    }.onFailure {
        Log.w(TAG, "tree serialization failed: ${it.message}")
    }.getOrNull()

    private fun serializeNode(node: NativeUINode): JSONObject {
        val objectValue = JSONObject()
            .put("id", node.id.toLong() and 0xffffffffL)
            .put("type", node.type)
            .put("on_press", node.onPress)
            .put("on_long_press", node.onLongPress)
            .put("props", serializeProps(node.props))

        node.layout?.let { objectValue.put("layout", serializeLayout(it)) }
        node.style?.let { objectValue.put("style", serializeStyle(it)) }

        val children = JSONArray()
        node.children.forEach { children.put(serializeNode(it)) }
        objectValue.put("children", children)

        return objectValue
    }

    private fun serializeLayout(layout: NodeLayout): JSONObject = JSONObject()
        .put("width", layout.width.toDouble())
        .put("width_mode", layout.widthMode)
        .put("height", layout.height.toDouble())
        .put("height_mode", layout.heightMode)
        .put("padding_top", layout.paddingTop.toDouble())
        .put("padding_right", layout.paddingRight.toDouble())
        .put("padding_bottom", layout.paddingBottom.toDouble())
        .put("padding_left", layout.paddingLeft.toDouble())
        .put("margin_top", layout.marginTop.toDouble())
        .put("margin_right", layout.marginRight.toDouble())
        .put("margin_bottom", layout.marginBottom.toDouble())
        .put("margin_left", layout.marginLeft.toDouble())
        .put("flex_grow", layout.flexGrow.toDouble())
        .put("flex_shrink", layout.flexShrink.toDouble())
        .put("flex_basis", layout.flexBasis.toDouble())
        .put("flex_direction", layout.flexDirection)
        .put("flex_wrap", layout.flexWrap)
        .put("align_self", layout.alignSelf)
        .put("align_items", layout.alignItems)
        .put("align_content", layout.alignContent)
        .put("justify_content", layout.justifyContent)
        .put("gap", layout.gap.toDouble())
        .put("row_gap", layout.rowGap.toDouble())
        .put("position_type", layout.positionType)
        .put("display", layout.display)
        .put("overflow", layout.overflow)
        .put("safe_area", layout.safeArea)

    private fun serializeStyle(style: NodeStyle): JSONObject = JSONObject()
        .put("bg_color", style.bgColor)
        .put("border_radius", style.borderRadius.toDouble())
        .put("border_width", style.borderWidth.toDouble())
        .put("border_color", style.borderColor)
        .put("opacity", style.opacity.toDouble())
        .put("elevation", style.elevation.toDouble())

    private fun serializeProps(props: GenericProps): JSONObject {
        val objectValue = JSONObject()
        for ((key, value) in props.entries) {
            runCatching { objectValue.put(key, if (value is List<*>) JSONArray(value) else value) }
        }
        return objectValue
    }
}

File: resources/xcode/NativePHP/NativeRender/NativeElementObservationRegistry.swift
import Foundation

/// Opt-in observation seam for tooling that consumes rendered trees or native
/// interaction events. Serialization is skipped while no observers exist.
final class NativeElementObservationRegistry {
    static let shared = NativeElementObservationRegistry()

    struct Subscription: Hashable {
        fileprivate enum Stream: Hashable { case trees, events }
        fileprivate let stream: Stream
        fileprivate let id: Int
    }

    private let lock = NSLock()
    private var sequence = 0
    private var treeObservers: [Int: (String) -> Void] = [:]
    private var eventObservers: [Int: (String) -> Void] = [:]
    private var latestTree: NativeUITree?

    private init() {}

    func observeTrees(_ observer: @escaping (String) -> Void) -> Subscription {
        lock.lock()
        sequence &+= 1
        let subscription = Subscription(stream: .trees, id: sequence)
        treeObservers[sequence] = observer
        let replay = latestTree
        lock.unlock()

        if let replay, let json = serializeTree(replay) {
            observer(json)
        }
        return subscription
    }

    func observeEvents(_ observer: @escaping (String) -> Void) -> Subscription {
        subscribe(stream: .events, observer: observer)
    }

    func unsubscribe(_ subscription: Subscription) {
        lock.lock(); defer { lock.unlock() }
        switch subscription.stream {
        case .trees:
            treeObservers.removeValue(forKey: subscription.id)
        case .events:
            eventObservers.removeValue(forKey: subscription.id)
        }
    }

    func publish(tree: NativeUITree) {
        lock.lock()
        latestTree = tree
        let observers = Array(treeObservers.values)
        lock.unlock()
        guard !observers.isEmpty, let json = serializeTree(tree) else { return }
        notify(observers, json: json)
    }

    func publishEvent(type: Int, callbackId: Int, nodeId: Int) {
        guard type <= EventType.sheetDismiss else { return }
        let observers = snapshot(stream: .events)
        guard !observers.isEmpty else { return }

        let payload: [String: Any] = [
            "type": type,
            "callback_id": callbackId,
            "node_id": nodeId,
            "timestamp": Int(Date().timeIntervalSince1970 * 1000),
        ]
        guard let data = try? JSONSerialization.data(withJSONObject: payload),
              let json = String(data: data, encoding: .utf8) else { return }
        notify(observers, json: json)
    }

    private func subscribe(
        stream: Subscription.Stream,
        observer: @escaping (String) -> Void
    ) -> Subscription {
        lock.lock(); defer { lock.unlock() }
        sequence &+= 1
        switch stream {
        case .trees:
            treeObservers[sequence] = observer
        case .events:
            eventObservers[sequence] = observer
        }
        return Subscription(stream: stream, id: sequence)
    }

    private func snapshot(stream: Subscription.Stream) -> [(String) -> Void] {
        lock.lock(); defer { lock.unlock() }
        switch stream {
        case .trees: return Array(treeObservers.values)
        case .events: return Array(eventObservers.values)
        }
    }

    private func notify(_ observers: [(String) -> Void], json: String) {
        for observer in observers {
            observer(json)
        }
    }

    func serializeTree(_ tree: NativeUITree) -> String? {
        let payload: [String: Any] = [
            "version": tree.version,
            "callback_count": tree.callbackCount,
            "root": serializeNode(tree.root),
        ]
        guard let data = try? JSONSerialization.data(withJSONObject: payload) else { return nil }
        return String(data: data, encoding: .utf8)
    }

    private func serializeNode(_ node: NativeUINode) -> [String: Any] {
        var object: [String: Any] = [
            "id": node.id,
            "type": node.type,
            "on_press": node.onPress,
            "on_long_press": node.onLongPress,
            "props": serializeProps(node.props),
            "children": node.children.map { serializeNode($0) },
        ]
        if let layout = node.layout { object["layout"] = serializeLayout(layout) }
        if let style = node.style { object["style"] = serializeStyle(style) }
        return object
    }

    private func serializeLayout(_ layout: NodeLayout) -> [String: Any] {
        [
            "width": Double(layout.width), "width_mode": layout.widthMode,
            "height": Double(layout.height), "height_mode": layout.heightMode,
            "padding_top": Double(layout.paddingTop), "padding_right": Double(layout.paddingRight),
            "padding_bottom": Double(layout.paddingBottom), "padding_left": Double(layout.paddingLeft),
            "margin_top": Double(layout.marginTop), "margin_right": Double(layout.marginRight),
            "margin_bottom": Double(layout.marginBottom), "margin_left": Double(layout.marginLeft),
            "flex_grow": Double(layout.flexGrow), "flex_shrink": Double(layout.flexShrink),
            "flex_basis": Double(layout.flexBasis), "flex_direction": layout.flexDirection,
            "flex_wrap": layout.flexWrap, "align_self": layout.alignSelf,
            "align_items": layout.alignItems, "align_content": layout.alignContent,
            "justify_content": layout.justifyContent, "gap": Double(layout.gap),
            "row_gap": Double(layout.rowGap), "position_type": layout.positionType,
            "display": layout.display, "overflow": layout.overflow, "safe_area": layout.safeArea,
        ]
    }

    private func serializeStyle(_ style: NodeStyle) -> [String: Any] {
        [
            "bg_color": style.bgColor, "border_radius": Double(style.borderRadius),
            "border_width": Double(style.borderWidth), "border_color": style.borderColor,
            "opacity": Double(style.opacity), "elevation": Double(style.elevation),
        ]
    }

    private func serializeProps(_ props: GenericProps) -> [String: Any] {
        var object: [String: Any] = [:]
        for (key, value) in props.entries {
            switch value {
            case let string as String: object[key] = string
            case let number as NSNumber: object[key] = number
            case let bool as Bool: object[key] = bool
            case let list as [String]: object[key] = list
            default: break
            }
        }
        return object
    }
}
